### PR TITLE
Alex dev

### PR DESCRIPTION
**Pull request by AlexanderFengler**
First attempt at publishable NN extension

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -9,7 +9,7 @@ Introduction
 :Mailing list: https://groups.google.com/group/hddm-users/
 :Copyright: This document has been placed in the public domain.
 :License: HDDM is released under the BSD 2 license.
-:Version: 0.7.8
+:Version: 0.9.0
```
Comment (twiecki) [inline README.rst:12]: Why not `0.8.0`?

File: hddm/cnn/config.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        self.inference_dataset = glob.glob('{}/*'.format(param_recovery_folder))
+        
+        self.dataset_dir = os.path.join(self.model_name, 'training_data_binned_{}_nbins_{}_n_{}'.format(int(self.isBinned),self.nBins,self.nDatapoints))
+        self.refname = self.dataset_dir.replace('/','_')
+
+        self.train_tfrecords = self.refname+'_train.tfrecords'
+        self.val_tfrecords = self.refname+'_val.tfrecords'
+        self.test_tfrecords = self.refname+'_test.tfrecords'
+
+        """
```
Comment (twiecki) [inline hddm/cnn/config.py:76]: Comments should be with `#`.

File: hddm/cnn/config.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        Specify whether or not to treat gaussian errors as isotropic
+        This only matters for the "reverse" model
+        """
+        self.full_cov_matrix = False
+        self.model_suffix = 'full_cov' if self.full_cov_matrix else 'isotropic'
+
+        # Data configuration
+        self.results_dir = '/media/data_cifs/projects/prj_approx-bayes/projectABC/results/'
```
Comment (twiecki) [inline hddm/cnn/config.py:84]: Why are these hardcoded?

File: hddm/cnn/reverse_model.py [hunk 4]
```diff
@@ -0,0 +1,426 @@
...
+                    data.append(ip)
+                    labels.append(op)
+                    preds.append(pred)
+                    print('{} batches complete..'.format(len(errors)))
+            except tf.errors.OutOfRangeError:
+                print('Epoch limit reached!')
+            finally:
+                coord.request_stop()
+            coord.join(threads)
+    import ipdb; ipdb.set_trace()
```
Comment (twiecki) [inline hddm/cnn/reverse_model.py:390]: stray debug.

File: hddm/cnn/train_detector.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+import tensorflow as tf
+from .config import *
+from .tf_data_handler import inputs
+import numpy as np
+import tqdm, time
+import matplotlib.pyplot as plt
+import math
+from textwrap import wrap
+
+class cnn_model_struct:
```
Comment (twiecki) [inline hddm/cnn/train_detector.py:18]: classes should be `CamelCase`: `CNNModelStruct`

File: hddm/cnn/train_detector.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+
+        step = 0
+        start = time.time()
+        try:
+            while not coord.should_stop():
+                # train for a step
+                _, loss, softmax_outputs, tr_data, tr_labels = sess.run([train_step, kl_divergence_loss, y_conv, train_data, train_labels])
+                step+=1
+                if math.isnan(loss):
+                    import ipdb; ipdb.set_trace()
```
Comment (twiecki) [inline hddm/cnn/train_detector.py:258]: remove

File: hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint [hunk 7]
```diff
@@ -0,0 +1,2 @@
+model_checkpoint_path: "/media/data_cifs/projects/prj_approx-bayes/projectABC/models/ornstein_training_data_binned_1_nbins_512_n_100000/ornstein_351000.ckpt-351000"
```
Comment (twiecki) [inline hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint:1]: all these paths are hardcoded, should checkpoints even be here?

File: hddm/likelihoods_cnn.py [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+                    data[:, cnt] = params[tmp_str].loc[value['rt'].index].values[:, 0]
+                    if (data[:, cnt].min() < model_config[model]['param_bounds'][0][cnt]) or (data[:, cnt].max() > model_config[model]['param_bounds'][1][cnt]):
+                        print('boundary violation of regressor part')
+                        return - np.inf
+                else:
+                    data[:, cnt] = params[tmp_str]
+
+                cnt += 1
+
+            # THIS IS NOT YET FINISHED !
```
Comment (twiecki) [inline hddm/likelihoods_cnn.py:400]: Shouldn't it be? :)

File: hddm/likelihoods_cnn.py [hunk 9]
```diff
@@ -0,0 +1,559 @@
...
+            # THIS IS NOT YET FINISHED !
```
Comment (twiecki) [inline hddm/likelihoods_cnn.py:439]: Bit unsettling :)

File: hddm/models/hddm_info.py [hunk 10]
```diff
@@ -100,23 +103,416 @@ class HDDM(HDDMBase):
...
+                             'g': 0.5}
+
         self.emcee_dispersions = {'a':1, 't': 0.1, 'a_std': 1, 't_std': 0.15, 'sz': 1.1, 'v': 1.5,
                                   'st': 0.1, 'sv': 3, 'z_trans': 0.2, 'z': 0.1,
                                   'p_outlier':1., 'v_std': 1, 'alpha': 1.5, 'pos_alpha': 1.5}
 
-
-        self.is_informative = kwargs.pop('informative', True)
+        
+        if hasattr(self, 'is_informative'):
```
Comment (twiecki) [inline hddm/models/hddm_info.py:133]: Why is this required?

File: hddm/models/hddm_info.py [hunk 11]
```diff
@@ -100,23 +103,416 @@ class HDDM(HDDMBase):
...
+            pass
+        else:
+            self.is_informative = kwargs.pop('informative', True)
+        
+        #self.is_informative = kwargs.pop('informative', True)
 
         super(HDDM, self).__init__(*args, **kwargs)
 
+        # AF ADDED ----------------------------------------------------------------------------
+        print('Is Neural Net? ', self.nn)
```
Comment (twiecki) [inline hddm/models/hddm_info.py:143]: These print statements will become annoying quite quickly.

File: hddm/models/hddm_info.py [hunk 12]
```diff
@@ -100,23 +103,416 @@ class HDDM(HDDMBase):
     """
 
     def __init__(self, *args, **kwargs):
-        self.slice_widths = {'a':1, 't':0.01, 'a_std': 1, 't_std': 0.15, 'sz': 1.1, 'v': 1.5,
-                             'st': 0.1, 'sv': 3, 'z_trans': 0.2, 'z': 0.1,
-                             'p_outlier':1., 'v_std': 1, 'alpha': 1.5, 'pos_alpha': 1.5}
+        # self.slice_widths = {'a':1, 't':0.01, 'a_std': 1, 't_std': 0.15, 'sz': 1.1, 'v': 1.5,
```
Comment (twiecki) [inline hddm/models/hddm_info.py:106]: I think you should create a new class for this and not change `HDDInfo`.

File: hddm/models/hddm_nn.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+            d['network'] = load_mlp(model = d['model'])
+            likelihood_ = hddm.likelihoods_mlp.make_mlp_likelihood(model = d['model'])
+            network_dict = {'network': d['network']}
+
+        d['wfpt_nn'] = stochastic_from_dist('Wienernn' + '_' + d['model'],
+                                            partial(likelihood_, **network_dict))
+        super(HDDMnn, self).__setstate__(d)
+
+
+# net_out = self.network(np.zeros((10, 1, 4, 1), dtype = np.float32))
```
Comment (twiecki) [inline hddm/models/hddm_nn.py:187]: Please remove all commented code.

File: hddm/models/hddm_nn_regression.py [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+
+            return pd.DataFrame(predictor, index = data.index)
+
+        return self.pymc_node(func, 
+                              kwargs['doc'], 
+                              name, 
+                              parents = parents,
+                              trace = self.keep_regressor_trace)
+
+class HDDMnnRegressor(HDDM):
```
Comment (twiecki) [inline hddm/models/hddm_nn_regression.py:72]: Can't you inherit from `HDDMRegressor` rather than copying it?

File: hddm/models/hddm_stimcoding_nn.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+from kabuki.utils import stochastic_from_dist
+from hddm.models import HDDM
+from hddm.keras_models import load_mlp
+from hddm.cnn.wrapper import load_cnn
+import hddm
+
+import wfpt
+from functools import partial
+
+class HDDMnnStimCoding(HDDM):
```
Comment (twiecki) [inline hddm/models/hddm_stimcoding_nn.py:15]: Also better to inherit here rather than copy.

File: hddm/models/hddm_stimcoding_nn.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+        return KnodeWfptStimCoding(self.wfpt_nn, 
+                                   'wfpt', # TD: ADD wfpt class we need
+                                   observed = True, 
+                                   col_name = ['rt', 'response'], # Note: This is different from vanilla stimcoding class where it was set to col_name = 'rt',
+                                   depends = [self.stim_col],
+                                   split_param = self.split_param,
+                                   stims = self.stims,
+                                   stim_col = self.stim_col,
+                                   **wfpt_parents)
+class KnodeWfptStimCoding(Knode):
```
Comment (twiecki) [inline hddm/models/hddm_stimcoding_nn.py:155]: Doesn't this already exist so that you can import it? Trying to avoid code duplication as much as possible.

File: hddm/simulators/hddm_dataset_generators.py [hunk 17]
```diff
@@ -0,0 +1,1265 @@
...
+            cnt += 1
+    else:
+        parameter_data = np.random.uniform(low = model_config[model]['param_bounds'][0],
+                                           high = model_config[model]['param_bounds'][1],
+                                           size = (n_parameter_vectors, len(model_config[model]['params'])))
+                                           
+    return pd.DataFrame(parameter_data, columns = model_config[model]['params'])
+
+# Dataset generators
+def simulator_single_subject(parameters = [0, 0, 0],
```
Comment (twiecki) [inline hddm/simulators/hddm_dataset_generators.py:167]: kwargs shouldn't be mutable: `(0, 0, 0)`.

File: hddm/simulators/hddm_dataset_generators.py [hunk 18]
```diff
@@ -0,0 +1,1265 @@
...
+    data_out.reset_index(drop = True, inplace = True)
+
+    if bin_pointwise:
+        data_out['rt'] = data_out['rt'].astype(np.int_)
+        data_out['response'] = data_out['response'].astype(np.int_)
+        #data_out['nn_response'] = data_out['nn_response'].astype(np.int_)
+
+    return (data_out, gt)
+
+def simulator_covariate(dependent_params = ['v'],
```
Comment (twiecki) [inline hddm/simulators/hddm_dataset_generators.py:518]:
```suggestion
def simulator_covariate(dependent_params = ('v',),
```

File: hddm/utils.py [hunk 19]
```diff
@@ -534,26 +578,27 @@ def qp_plot(x, groupby=None, quantiles=(0.1, 0.3, 0.5, 0.7, 0.9), ncols=None, dr
 
     return ax
 
+    # AF-Q: UNUSED?
     #create axes for subjects
-    n_subjs = len(data.subj_idx.unique())
```
Comment (twiecki) [inline hddm/utils.py:538]: this happens after the `return`? Then yeah, remove.

File: requirements.txt [hunk 20]
```diff
@@ -1,12 +1,19 @@
...
+pymc>=2.3.8
 python-dateutil==1.5
 pytz==2012d
 wsgiref==0.1.2
 matplotlib>=1.1.0
 git+https://github.com/hddm-devs/kabuki.git@master
-patsy
+seaborn>=0.11.0 
+statsmodels>=0.12.0
+tensorflow==1.15
```
Comment (twiecki) [inline requirements.txt:16]: Does it need to be this specific?

**Comment by twiecki** [comment]
Well, it's not going to win any coding awards ;) but holy cow, a *lot* of work went into this. Really cool new features. No way I'll be able to review but I left some stylistic points of feedback that I think should be addressed. Adding `tf` as a dependency is quite heavy, have you considered making that option for people who want to use this new functionality?

File: hddm/cnn/config.py (hunk 3 above)
Comment (AlexanderFengler) [reply hddm/cnn/config.py:84]: Ah no that's right, these paths are actually all irrelevant. 
As a global comment on the cnn code --> the cnn part is not emphasized because it has performance issues relative to the mlp stuff. The code wasn't fully cleaned up, as I realize now.

Comment (twiecki) [reply hddm/cnn/config.py:84]: I would add only the minimum that is necessary here, and only the best bits, the more code there is to maintain, the more headache.

File: requirements.txt (hunk 20 above)
Comment (AlexanderFengler) [reply requirements.txt:16]: tensorflow is fixed at 1.15 because of some issues regarding performance when switching to tf 2.x. 
Recently an update h5py also affected saved model parsing (which was sort of fixed with newest tf 2.5).
To relax the req, we need to do some more testing to make ensure it's working regardless of version.

seaborn and statsmodels I can probably drop the version requirement to earlier if that's desirable.

File: hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint (hunk 7 above)
Comment (AlexanderFengler) [reply hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint:1]: when the class is initialized its overwritten to the correct checkpoint path for the user. 
But I guess I can find a better solution for this one.

File: hddm/cnn/config.py (hunk 2 above)
Comment (AlexanderFengler) [reply hddm/cnn/config.py]: fixed

File: hddm/cnn/config.py (hunk 3 above)
Comment (AlexanderFengler) [reply hddm/cnn/config.py]: severely reduced the cnn related code

File: hddm/cnn/reverse_model.py (hunk 4 above)
Comment (AlexanderFengler) [reply hddm/cnn/reverse_model.py]: done

File: hddm/cnn/train_detector.py (hunk 5 above)
Comment (AlexanderFengler) [reply hddm/cnn/train_detector.py]: done

File: hddm/cnn/train_detector.py (hunk 6 above)
Comment (AlexanderFengler) [reply hddm/cnn/train_detector.py]: done

File: hddm/likelihoods_cnn.py (hunk 8 above)
Comment (AlexanderFengler) [reply hddm/likelihoods_cnn.py]: made comment less frightening

File: hddm/likelihoods_cnn.py (hunk 9 above)
Comment (AlexanderFengler) [reply hddm/likelihoods_cnn.py]: as above

File: hddm/models/hddm_info.py (hunk 11 above)
Comment (AlexanderFengler) [reply hddm/models/hddm_info.py]: reduced print statements

File: hddm/models/hddm_nn.py (hunk 13 above)
Comment (AlexanderFengler) [reply hddm/models/hddm_nn.py]: done

File: hddm/models/hddm_nn_regression.py (hunk 14 above)
Comment (AlexanderFengler) [reply hddm/models/hddm_nn_regression.py]: done

File: hddm/models/hddm_stimcoding_nn.py (hunk 15 above)
Comment (AlexanderFengler) [reply hddm/models/hddm_stimcoding_nn.py]: done

File: hddm/utils.py (hunk 19 above)
Comment (AlexanderFengler) [reply hddm/utils.py:538]: done

File: hddm/models/hddm_info.py (hunk 10 above)
Comment (AlexanderFengler) [reply hddm/models/hddm_info.py]: fixed

File: hddm/models/hddm_info.py (hunk 12 above)
Comment (AlexanderFengler) [reply hddm/models/hddm_info.py]: This would be the only point where I would disagree. Maybe serious refactoring could facilitate that, but without it, not changing HDDMInfo would rather lead to code duplication on the level of the various HDDMnn classes. This was changed also to accomodate HDDMrl (see 'alpha' and 'pos_alpha' additions to self.slice_widths for example). If you strongly disagree with this I will defer to you of course.

File: hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint (hunk 7 above)
Comment (AlexanderFengler) [reply hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint:1]: fixed this

File: HDDM.egg-info/dependency_links.txt [hunk 21]
```diff
@@ -0,0 +1 @@
+
```
Comment (twiecki) [inline HDDM.egg-info/dependency_links.txt:1]: These files shouldn't be here.

File: hddm/cnn/config.py [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+
+    def __init__(self, model=None, bins=None, N=None):
+        # Directory setup
+        self.base_dir = hddm.__path__[0]
+        #self.data_dir = 'data'
+        #self.tfrecord_dir = 'tfrecords'
+        #self.summary_dir = 'summaries'
+
+        # training dataset features
+        self.isBinned = True
```
Comment (twiecki) [inline hddm/cnn/config.py:14]:
```suggestion
        self.binned = True
```

File: hddm/cnn/config.py [hunk 23]
```diff
@@ -0,0 +1,229 @@
...
+        if model is None:
+            model = 'ddm_sdv'
+        if bins != None:
+            self.nBins = bins
```
Comment (twiecki) [inline hddm/cnn/config.py:18]: Everything but class names should be `snake_case`.

File: hddm/cnn/train_detector.py [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+                var = tf.get_variable(name=var_name, initializer=value)
+        else:
+            var = tf.constant(value, dtype=tf.float32, name=var_name)
+            #var = tf.get_variable(name=var_name, initializer=value)
+
+        self.var_dict[(name, idx)] = var
+
+        return var
+
+# def calc_error(labels,predictions):
```
Comment (twiecki) [inline hddm/cnn/train_detector.py:155]: Remove commented code.

File: hddm/cnn_models/angle_training_data_binned_1_nbins_512_n_100000/no_checkpoint [hunk 25]
```diff
@@ -0,0 +1,2 @@
+model_checkpoint_path: "/media/data_cifs/projects/prj_approx-bayes/projectABC/models/angle_training_data_binned_1_nbins_512_n_100000/angle_210500.ckpt-210500"
```
Comment (twiecki) [inline hddm/cnn_models/angle_training_data_binned_1_nbins_512_n_100000/no_checkpoint:1]: Why are these files required? They still have hardcoded paths.

File: hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint (hunk 7 above)
Comment (twiecki) [reply hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint:1]: Why not just remove them?

File: hddm/graphs.py [hunk 26]
```diff
@@ -0,0 +1,1538 @@
...
+from scipy.stats import truncnorm
+from scipy.stats import scoreatpercentile
+from scipy.stats.mstats import mquantiles
+from copy import deepcopy
+
+model_config = hddm.simulators.model_config
+
+# PREPROCESSING FUNCTIONS ------------------------------------------------------------------------------------------------------------------
+
+def untransform_traces(traces = None, model = None, is_nn = False):
```
Comment (twiecki) [inline hddm/graphs.py:27]: Please see pep8 for style guide.

File: hddm/graphs.py [hunk 27]
```diff
@@ -0,0 +1,1538 @@
...
+    """ 
+        Check which traces are transformed and apply inverse logit transform to them. (This works only if the transformed variable was
```
Comment (twiecki) [inline hddm/graphs.py:29]: Shouldn't be indented.

File: hddm/graphs.py [hunk 28]
```diff
@@ -0,0 +1,1538 @@
+from hddm.simulators import *
```
Comment (twiecki) [inline hddm/graphs.py:1]: `plotting.py` is the more common name.

File: hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint (hunk 7 above)
Comment (AlexanderFengler) [reply hddm/cnn_models/ornstein_training_data_binned_1_nbins_512_n_100000/checkpoint:1]: done

File: hddm/cnn_models/angle_training_data_binned_1_nbins_512_n_100000/no_checkpoint (hunk 25 above)
Comment (AlexanderFengler) [reply hddm/cnn_models/angle_training_data_binned_1_nbins_512_n_100000/no_checkpoint:1]: gone now.

File: HDDM.egg-info/PKG-INFO [hunk 29]
```diff
@@ -0,0 +1,17 @@
+Metadata-Version: 1.1
```
Comment (twiecki) [inline HDDM.egg-info/PKG-INFO:1]: This whole dir shouldn't exist.

File: hddm/cnn/config.py (hunk 22 above)
Comment (AlexanderFengler) [reply hddm/cnn/config.py:14]: fixed

File: hddm/cnn/config.py (hunk 23 above)
Comment (AlexanderFengler) [reply hddm/cnn/config.py:18]: ok I think I caught all of those. Convention was mostly applied.

File: hddm/graphs.py (hunk 26 above)
Comment (AlexanderFengler) [reply hddm/graphs.py]: applied black

File: hddm/graphs.py (hunk 27 above)
Comment (AlexanderFengler) [reply hddm/graphs.py]: blacked

File: hddm/graphs.py (hunk 28 above)
Comment (AlexanderFengler) [reply hddm/graphs.py]: renamed

File: HDDM.egg-info/PKG-INFO (hunk 29 above)
Comment (AlexanderFengler) [reply HDDM.egg-info/PKG-INFO]: gone. Not sure why this was still there since it had been included in .gitignore already

**Comment by twiecki** [comment]
Did you run black on the full code-base?

**Comment by AlexanderFengler** [comment]
Ran it on the hddm folder yeah. I thought it doesn't makes sense to just run it once over everything for consistency.

File: hddm/cnn/train_detector.py (hunk 24 above)
Comment (AlexanderFengler) [reply hddm/cnn/train_detector.py]: done

**Comment by twiecki** [comment]
Odd, I have run it too on the master branch and then tried to point this PR to it but there were conflicts.

**Comment by twiecki** [comment]
OK, I resolved the conflicts, so you have to pull the changes from this branch.

**Comment by twiecki** [comment]
Not sure why but I had to resolve conflicts again, so make sure to pull.

**Comment by AlexanderFengler** [comment]
Hm maybe I am doing something wrong there, but I pulled before pushing again.
